### ISSUE TITLE
C++11: thrust::result_of  fails if parameters are passed by non-const reference

### ISSUE THREAD

**Issue by dachziegel** [thrust/thrust#678]
The following example fails to compile with `-std=c++11`.
When `operator()(T& t)` is changed to `operator()(T t)` or `operator()(const T& t)` it compiles.

    #include <thrust/device_vector.h>
    #include <thrust/tuple.h>
    #include <thrust/iterator/transform_iterator.h>

    struct test_cpp11
    {
      template <typename T>
      T operator()(T& t)
      {
        return t;
      }
    };

    int main()
    {
      thrust::device_vector<int> v1(10);
      thrust::device_vector<int> v2(10);
      
      auto b = thrust::make_transform_iterator(thrust::make_zip_iterator(thrust::make_tuple(v1.begin(), v2.begin())), test_cpp11());
      auto e = thrust::make_transform_iterator(thrust::make_zip_iterator(thrust::make_tuple(v1.end(), v2.end())), test_cpp11());
      
      thrust::copy(b, e, thrust::make_zip_iterator(thrust::make_tuple(v1.begin(), v2.begin())));
    }

**error messages**

    In file included from result_of_fail.cpp:1:
    In file included from /opt/thrust-git/thrust/device_vector.h:26:
    In file included from /opt/thrust-git/thrust/detail/vector_base.h:29:
    In file included from /opt/thrust-git/thrust/detail/contiguous_storage.h:155:
    In file included from /opt/thrust-git/thrust/detail/contiguous_storage.inl:22:
    In file included from /opt/thrust-git/thrust/detail/allocator/copy_construct_range.h:46:
    In file included from /opt/thrust-git/thrust/detail/allocator/copy_construct_range.inl:21:
    In file included from /opt/thrust-git/thrust/detail/copy.h:90:
    In file included from /opt/thrust-git/thrust/detail/copy.inl:21:
    In file included from /opt/thrust-git/thrust/system/detail/generic/copy.h:58:
    In file included from /opt/thrust-git/thrust/system/detail/generic/copy.inl:21:
    In file included from /opt/thrust-git/thrust/functional.h:26:
    In file included from /opt/thrust-git/thrust/detail/functional/placeholder.h:20:
    In file included from /opt/thrust-git/thrust/detail/functional/actor.h:33:
    /opt/thrust-git/thrust/detail/type_traits/result_of.h:38:47: error: no type named 'type' in 'std::result_of<test_cpp11 (thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>)>'
      typedef typename std::result_of<Signature>::type type;
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
    /opt/thrust-git/thrust/detail/type_traits.h:413:20: note: in instantiation of template class 'thrust::detail::result_of<test_cpp11 (thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>)>' requested here
      typedef typename Then::type type;
                      ^
    /opt/thrust-git/thrust/iterator/detail/iterator_adaptor_base.h:48:5: note: in instantiation of template class 'thrust::detail::eval_if<true, thrust::detail::result_of<test_cpp11 (thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>)>, thrust::detail::identity_<thrust::use_default> >' requested here
      : thrust::detail::eval_if<
        ^
    /opt/thrust-git/thrust/iterator/detail/transform_iterator.inl:38:38: note: in instantiation of template class 'thrust::detail::ia_dflt_help<thrust::use_default, thrust::detail::result_of<test_cpp11 (thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>)> >' requested here
        typedef typename thrust::detail::ia_dflt_help<
                                        ^
    /opt/thrust-git/thrust/iterator/transform_iterator.h:191:22: note: in instantiation of template class 'thrust::detail::transform_iterator_base<test_cpp11, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::use_default, thrust::use_default>' requested here
        : public detail::transform_iterator_base<AdaptableUnaryFunction, Iterator, Reference, Value>::type
                        ^
    result_of_fail.cpp:21:12: note: in instantiation of template class 'thrust::transform_iterator<test_cpp11, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::use_default, thrust::use_default>' requested here
      auto b = thrust::make_transform_iterator(thrust::make_zip_iterator(thrust::make_tuple(v1.begin(), v2.begin())), test_cpp11());
              ^
    In file included from result_of_fail.cpp:1:
    In file included from /opt/thrust-git/thrust/device_vector.h:26:
    In file included from /opt/thrust-git/thrust/detail/vector_base.h:533:
    In file included from /opt/thrust-git/thrust/detail/vector_base.inl:25:
    In file included from /opt/thrust-git/thrust/equal.h:237:
    In file included from /opt/thrust-git/thrust/detail/equal.inl:25:
    In file included from /opt/thrust-git/thrust/system/detail/generic/equal.h:47:
    In file included from /opt/thrust-git/thrust/system/detail/generic/equal.inl:21:
    In file included from /opt/thrust-git/thrust/mismatch.h:259:
    In file included from /opt/thrust-git/thrust/detail/mismatch.inl:27:
    In file included from /opt/thrust-git/thrust/system/detail/generic/mismatch.h:57:
    In file included from /opt/thrust-git/thrust/system/detail/generic/mismatch.inl:21:
    In file included from /opt/thrust-git/thrust/find.h:384:
    In file included from /opt/thrust-git/thrust/detail/find.inl:25:
    In file included from /opt/thrust-git/thrust/system/detail/generic/find.h:62:
    In file included from /opt/thrust-git/thrust/system/detail/generic/find.inl:24:
    /opt/thrust-git/thrust/iterator/transform_iterator.h:218:9: error: type 'super_t' (aka 'iterator_adaptor<transform_iterator<test_cpp11, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::use_default, thrust::use_default>, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, cv_value_type, thrust::use_default, typename thrust::iterator_traits<zip_iterator<tuple<normal_iterator<device_ptr<int> >, normal_iterator<device_ptr<int> >, null_type, null_type, null_type, null_type, null_type, null_type, null_type, null_type> > >::iterator_category, reference>') is not a direct or virtual base of 'thrust::transform_iterator<test_cpp11, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::use_default, thrust::use_default>'
          : super_t(x), m_f(f) {
            ^~~~~~~
    /opt/thrust-git/thrust/iterator/transform_iterator.h:334:10: note: in instantiation of member function 'thrust::transform_iterator<test_cpp11, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::use_default, thrust::use_default>::transform_iterator' requested here
      return transform_iterator<AdaptableUnaryFunction, Iterator>(it, fun);
            ^
    result_of_fail.cpp:21:20: note: in instantiation of function template specialization 'thrust::make_transform_iterator<test_cpp11, thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::detail::normal_iterator<thrust::device_ptr<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> > >' requested here
      auto b = thrust::make_transform_iterator(thrust::make_zip_iterator(thrust::make_tuple(v1.begin(), v2.begin())), test_cpp11());
                      ^
    2 errors generated.

**Comment by jaredhoberock** [comment]
I don't think this is an error -- zip_iterators don't return references, so there's no way it can bind to the `T&` expected by `test_cpp11`.

**Comment by dachziegel** [comment]
However, it compiles without c++11:

    #include <thrust/device_vector.h>
    #include <thrust/tuple.h>
    #include <thrust/iterator/transform_iterator.h>

    template <typename T>
    struct test_cpp03 : public thrust::unary_function<T&,T>
    {
      T operator()(T& t)
      {
        return t;
      }
    };

    int main()
    {
      thrust::device_vector<int> v1(10);
      thrust::device_vector<int> v2(10);

      thrust::copy(thrust::make_transform_iterator(thrust::make_zip_iterator(thrust::make_tuple(v1.begin(), v2.begin())), test_cpp03<thrust::tuple<int,int> >()),
                  thrust::make_transform_iterator(thrust::make_zip_iterator(thrust::make_tuple(v1.end(), v2.end())), test_cpp03<thrust::tuple<int,int> >()),
                  thrust::make_zip_iterator(thrust::make_tuple(v1.begin(), v2.begin())));
    }

**Comment by jaredhoberock** [comment]
That's because you're misrepresenting the argument to `unary_function` with the `T&`. zip_iterator returns a temporary, not a reference to some named tuple on the stack or some tuple on the heap.

**Comment by dachziegel** [comment]
But it also compiles with

    struct test_cpp03 : public thrust::unary_function<T,T>

**Comment by jaredhoberock** [comment]
Then use `unary_function` :-)

**Comment by dachziegel** [comment]
I have a fix for this issue ready:

    #if __cplusplus >= 201103L || (defined(__cpp_variadic_templates) && defined(__cpp_lib_result_of_sfinae))

    template<typename Signature>
      struct result_of;
      
    template<typename R, typename... Args>
      struct result_of<R(Args...)>
    {
      typedef typename std::result_of<R(Args&...)>::type type;
    };

    #else

**Comment by jaredhoberock** [comment]
I don't think that implementing `result_of` incorrectly is a good way to resolve this issue. It's just a workaround that forces your use case to compile.

You might investigate what happens if [`result_of` is applied to `iterator_reference` rather than `iterator_value`](https://github.com/thrust/thrust/blob/master/thrust/iterator/detail/transform_iterator.inl#L40).

**Comment by jaredhoberock** [comment]
Marking this as an enhancement -- I believe the compiler error is correct, but don't know at the moment how to relax what `transform_iterator` + `zip_iterator` do.

**Comment by jaredhoberock** [comment]
I think the right way to resolve this in a way that ensures that c++11 mode does not break existing Thrust programs is to have `transform_iterator` use something like `result_of_adaptable_function` rather than `result_of` directly. The difference would be that `result_of_adaptable_function` would check the user's function for the existence of a member type named `result_type`. If such a type exists, it would use this type instead of deducing the `result_type` from `result_of`.

This wouldn't cause the program in the OP to compile correctly in the absence of `unary_function`, but it would allow for interoperation between `unary_function` & c++11 as a workaround.

**Comment by dachziegel** [comment]
how about this:

    #if __cplusplus >= 201103L || (defined(__cpp_variadic_templates) && defined(__cpp_lib_result_of_sfinae))

    template<typename Functor, typename Args...>
      struct result_of<Functor(Args...), std::enable_if<!std::is_base_of<thrust::unary_function, Functor>::value
                                                        &&
                                                        !std::is_base_of<thrust::binary_function, Functor>::value
                                                        >::type>
    {
      typedef typename std::result_of<Signature>::type type;
    };

    #endif // __cplusplus >= 201103L

Only if C++11 is active **and** the functor does not inherit from either `thrust::unary_function` or `thrust::binary_function`, `std::result_of`is used to deduce the return type.

**Comment by jaredhoberock** [comment]
I believe a comprehensive solution must check for the existence of `result_type`, not the existence of a particular base class.

For example, the solution you propose would fail if the user's functor inherited from `std::unary_function`, or if the user's functor manually defined a typedef named `result_type`.

Moreover, we must not change the meaning of `result_of` in c++11 mode -- it should map directly to `std::result_of`.

**Comment by dachziegel** [comment]
I had just modified my comment to check for `result_type`.

**Comment by jaredhoberock** [comment]
> What do you mean by "map directly to std::result_of"?

I just mean that if there is a facility within Thrust named `result_of`, it must be identical in meaning to `std::result_of`. The use of `enable_if` in your solution changes the meaning to search for `result_type`, but `std::result_of` doesn't do that.

Anyway, your updated solution seems almost correct to me, but it needs a different name. I'd suggest `result_of_adaptable_function`.

To be complete, I think `result_of_adaptable_function` should first check for the existence of a member type named `result_type`, and if one does not exist, it should use `result_of`. I don't think our current implementation of `result_of` needs to be changed at all.

Finally, to implement the change, we should do a global search & replace within Thrust to change the use of `result_of` to `result_of_adaptable_function`.

**Comment by dachziegel** [comment]
Just so I understand you right: 

1. we rename `result_of` to `result_of_adaptable_function` everywhere, even within `result_of.h`?

2. *"To be complete, I think result_of_adaptable_function should first check for the existence of a member type named result_type, and if one does not exist, it should use result_of"*
This would be my `enable_if` based solution I posted above, right?

**Comment by jaredhoberock** [comment]
> we rename result_of to result_of_adaptable_function everywhere, even within result_of.h?

Not within result_of.h -- keep the definition of `result_of` as it appears in result_of.h the same. I'm suggesting adding new functionality named `result_of_adaptable_function` in addition to `result_of`.

In fact, since `thrust::detail::result_of` would not add anything at this point, I'd suggest removing it in favor of only having `result_of_adaptable_function`. So in reality I think I'm just suggesting renaming this functionality to avoid confusion.

> This would be my enable_if based solution I posted above, right?

Correct, but I'm suggesting something more along the lines of:

```
template<typename Functor, typename... Args>
  struct result_of_adaptable_function<
    Functor(Args...),
    typename  thrust::detail::enable_if<!thrust::detail::has_result_type<Functor>::value>::type
  > : thrust::detail::result_of<Functor(Args...)> {};
```

I think I'd also find a way to express it in c++03 rather than c++11.

**Comment by dachziegel** [comment]
Ok, now I am confused. 

1.
> Not within result_of.h -- keep the definition of result_of as it appears in result_of.h the same. I'm suggesting adding new functionality named result_of_adaptable_function in addition to result_of

-> `result_of` is still there with its old name in addition to `result_of_adaptable_function`

2.
>In fact, since thrust::detail::result_of would not add anything at this point, I'd suggest removing it in favor of only having result_of_adaptable_function. So in reality I think I'm just suggesting renaming this functionality to avoid confusion.

-> `result_of` is replaced by `result_of_adaptable_function`, `result_of_adaptable_function` adds the check for `return_type`.

which of these two paths do you want to take?

**Comment by jaredhoberock** [comment]
I've been referring to what exists in the Thrust master branch.

Let's just get rid of `thrust::detail::result_of` in favor of `thrust::detail::result_of_adaptable_function`. `thrust::detail::result_of` has no reason to exist, as far as I can tell. It can't be implemented correctly in c++03 mode, and in c++11 mode, `std::result_of` is available.

In summary, the fix to this issue would:

1. Eliminate `thrust::detail::result_of`
2. Introduce `thrust::detail::result_of_adaptable_function` which would check the function for `result_type`. If it exists, return it. If it does not exist, and c++11 is available, it would then fall back to `std::result_of`. If it does not exist, and c++11 is not available, it's a compiler error.
3. Replace all use of `thrust::detail::result_of` with `thrust::detail::result_of_adaptable_function`.

Does this make sense? Thanks for your patience; it's important to get this right.

**Comment by dachziegel** [comment]
Yes I understood that now, I'll see into it.

**Comment by dachziegel** [comment]
can this issue be closed now that #728 is merged?
